src/bun.js/test/jest.zig

```diff
@@ -430,6 +493,213 @@ pub const TestRunner = struct {
+
+                const value_start = i;
+                if (value_start >= file_buf.items.len - 1) break :outer;
+
+                const value_end: usize = brk: {
+                    while (strings.indexOf(file_buf.items[i..], "`;")) |j| {
+                        i += j + "`;".len;
+                        if (file_buf.items[i - 3] != '\\') {
+                            break :brk i - 2;
+                        }
+                    }
+
+                    break :outer;
+                };
+
+                if (value_end > file_buf.items.len) break :outer;
+
+                const key = file_buf.items[key_start..key_end];
+                const pretty_value = try default_allocator.alloc(u8, value_end - value_start);
+                bun.copy(u8, pretty_value, file_buf.items[value_start..value_end]);
+                const name_hash = std.hash.Wyhash.hash(0, key);
+                try values.put(name_hash, pretty_value);
+            }
+        }
+
+        pub fn cleanup() !void {
+            if (snapshot_count == 0) return;
+
+            const dir_path = runner.?.files.get(current_file_id).source.path.name.dirWithTrailingSlash();
+            if (!strings.eql(dir_path, snapshot_dir_path)) {
+                snapshot_dir.close();
+                snapshot_dir = try getSnapshotDir(dir_path);
+            }
+
+            try snapshot_file.seekTo(0);
+            try snapshot_file.writeAll(file_buf.items);
+            file_buf.clearAndFree();
+
+            var value_itr = values.valueIterator();
+            while (value_itr.next()) |value| {
+                default_allocator.free(value.*);
+            }
+            values.deinit();
+
+            snapshot_file.close();
+
+            var count_itr = counts.iterator();
+            while (count_itr.next()) |entry| {
+                default_allocator.free(entry.key_ptr.*);
+            }
+            counts.clearAndFree();
+        }
+
+        pub fn getSnapshotFile(file_id: TestRunner.File.ID) !std.fs.File {
+            const test_file = Jest.runner.?.files.get(file_id);
+            const test_filename = test_file.source.path.name.filename;
+            const dir_path = test_file.source.path.name.dirWithTrailingSlash();
+
+            var snapshot_filename_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
+            bun.copy(u8, &snapshot_filename_buf, test_filename);
+            bun.copy(u8, snapshot_filename_buf[test_filename.len..], ".snap");
+            const snapshot_filename = snapshot_filename_buf[0 .. test_filename.len + ".snap".len];
+
+            if (!strings.eql(dir_path, snapshot_dir_path)) {
+                snapshot_dir = try getSnapshotDir(dir_path);
+            }
+
+            if (update_snapshots) {
+                const file = try snapshot_dir.createFile(snapshot_filename, .{ .read = true });
+                return file;
+            }
+
+            const file = snapshot_dir.openFile(snapshot_filename, .{ .mode = .read_write }) catch |err| {
```

Comment:
`JSC.Syscall.open`

Reply:
maybe add a openFile wrapper to JSC.Syscall

but TLDR

We will have much better error logging and no undefined behavior if we avoid std.os-related APIs for situations like this

Reply:
Switched `std.fs` makedir and openfile to `JSC.Syscall.mkdir` and `JSC.Syscall.open`. Moved most of the logic into `getSnapshotFile` because there wasn't a need to keep the dir fd anymore